File: Data Structure/Segment Tree/LET_no307_Range Sum Query - Mutable.py
class Node:
    def __init__(self, i, j):
        self.i, self.j = i, j
        self.left = self.right = None
        self.val = 0

class SegTree:
    def __init__(self, nums):
        self.root = self.build(nums, 0, len(nums) - 1)
        
    def build(self, nums, i, j):
        if i == j:
            cur = Node(i, j)
            cur.val = nums[i]
            return cur
        
        mid = (i + j) // 2
        cur = Node(i, j)
        cur.left = self.build(nums, i, mid)
        cur.right = self.build(nums, mid + 1, j)
        cur.val = cur.left.val + cur.right.val
        return cur
    
    def update(self, node, idx, val):
        if node.i == node.j:
            node.val = val
            return 
        
        mid = (node.i + node.j) // 2
        if idx <= mid:
            self.update(node.left, idx, val)
        else:
            self.update(node.right, idx, val)
        node.val = node.left.val + node.right.val
        
    def query(self, node, i, j):
        if node.i == i and node.j == j:
            return node.val
        
        mid = (node.i + node.j) // 2
        if j <= mid:
            return self.query(node.left, i, j)
        if i >= mid + 1:
            return self.query(node.right, i, j)
        return self.query(node.left, i, mid) + self.query(node.right, mid + 1, j)

class NumArray:

    def __init__(self, nums: List[int]):
        self.seg = SegTree(nums)

    def update(self, index: int, val: int) -> None:
        self.seg.update(self.seg.root, index, val)

    def sumRange(self, left: int, right: int) -> int:
        return self.seg.query(self.seg.root, left, right)


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)

File: Data Structure/Segment Tree/LET_no307_Range Sum Query - Mutable.cpp
struct Node {
    int i, j, val;
    struct Node* left;
    struct Node* right;
    Node(int i, int j) {
        this->i = i;
        this->j = j;
        left = right = nullptr;
        val = 0;
    }
};
typedef struct Node Node;

class SegTree {
private:    
    Node* build(vector<int>& nums, int i, int j) {
        if (i == j) {
            Node* cur = new Node(i, j);
            cur->val = nums[i];
            return cur;
        }

        int mid = i + (j - i) / 2;
        Node* cur = new Node(i, j);
        cur->left = build(nums, i, mid);
        cur->right = build(nums, mid + 1, j);
        cur->val = cur->left->val + cur->right->val;
        return cur;
    }

public:
    Node* root;
    
    SegTree(vector<int>& nums) {
        root = build(nums, 0, nums.size() - 1);
    }

    void update(Node* node, int idx, int val) {
        if (node->i == node->j) {
            node->val = val;
            return;
        }

        int mid = node->i + (node->j - node->i) / 2;
        if (idx <= mid)
            update(node->left, idx, val);
        else
            update(node->right, idx, val);
        node->val = node->left->val + node->right->val;
    }

    int query(Node* node, int i, int j) {
        if (node->i == i && node->j == j)
            return node->val;

        int mid = node->i + (node->j - node->i) / 2;
        if (j <= mid)
            return query(node->left, i, j);
        if (i >= mid + 1)
            return query(node->right, i, j);
        return query(node->left, i, mid) + query(node->right, mid + 1, j);
    }
};

class NumArray {
private:
    SegTree* seg;
public:
    NumArray(vector<int>& nums) {
        seg = new SegTree(nums);
    }
    
    void update(int index, int val) {
        seg->update(seg->root, index, val);
    }
    
    int sumRange(int left, int right) {
        return seg->query(seg->root, left, right);
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * obj->update(index,val);
 * int param_2 = obj->sumRange(left,right);
 */